### PR TITLE
Feature: editable captions/subtitles, and configurable max_resolution, mp4_support and encoding

### LINKED ISSUE #62: Feature Request: secure playback
Hi thanks for this Strapi plugin. The uploader is very nice.

I would love to see [secure video playback](https://docs.mux.com/guides/video/secure-video-playback) support in this plugin. My use case is a video sharing website with 20K viewers a week. I need to ensure that the mux video uploads are only watchable by logged in patrons.

I think this is doable right now using users-permissions plugin and limiting read access to mux assets to only users in a Patron role. However, this setup doesn't limit sharing the link with non-patrons which is something I need to do in order to keep cost managable.

cheers

### PR DESCRIPTION
**Pull request by hdoro**
This PR makes it possible to create assets with configurable `max_resolution`, `mp4_support` and `encoding_tier`, as well as uploading custom subtitles/captions, or requesting auto-generation.

Closes #62 #47 #72 #68

### CODE REVIEW

File: server/utils/textTracks.ts [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    text_type: 'subtitles',
+    closed_captions: track.closed_captions,
+    language_code: track.language_code,
+    name: track.closed_captions ? `${track.name} (CC)` : track.name,
+    url: getStrapiTextTrackUrl(track.id),
+  };
+}
+
+const BASE_URL = 'https://nice-falcon-11.telebit.io';
+// const BASE_URL = (strapi as any).backendURL
```
Comment (erikpena) [inline server/utils/textTracks.ts:31]: This looks like wasn't intended to stay

File: server/utils/textTracks.ts [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+  // Give mux' servers enough time to download tracks before deleting them
+  if (alreadyInMuxWithoutFile.length > 0) {
+    await sleep(1000);
+  }
+}
+
+export function sleep(ms: number) {
+  return new Promise((resolve) => {
+    setTimeout(resolve, ms);
+  });
```
Comment (erikpena) [inline server/utils/textTracks.ts:186]: This feels like it should be utilizing our webhooks to determine if the file was downloaded (or failed):

Webhook events I would use to determine this:
- https://docs.mux.com/webhook-reference#video.asset.track.ready
- https://docs.mux.com/webhook-reference#video.asset.track.errored

**Comment by erikpena** [comment]
Hello @hdoro,
Thank you for the PR for this.  I've taken some time to review this and have some notes that I want to share with you:

- Naming conventions: for this project, we are using kebab casing for filenames but it looks like you're using capital casing.  We'd want to switch to using kebab-casing if possible (e.g. `SignedTokensProvider.tsx` -> `signed-tokens-provider.tsx`).
- I am unable to build/test this plugin in a local environment due to some type issues.  I think it is related to my comment regarding upgrading the versions of react/react-dom and incompatibilities with the react-router-dom dependencies.

File: package.json [hunk 3]
```diff
@@ -23,43 +23,47 @@
     }
   ],
   "scripts": {
-    "build": "yarn clean && tsc && yarn copy-files",
+    "build": "npm run clean && tsc && npm run copy-files",
```
[the remark this answers is not in the document]
Comment (hdoro) [reply package.json:26]: I was struggling to get yarn v1 to run (I had v3 installed), so I switched to npm and forgot to undo it.

I could argue that npm is more generic and democratic as it's shipped with every NodeJS installation, and it'll also work on yarn installations, etc. etc. But instead I've simply rolled it back 😬

**Comment by hdoro** [comment]
@erikpena thank you for the thorough review!

I've went through every item and created [a repository configured for testing the plugin](https://github.com/hdoro/strapi-mux-test), which is documented for how to run it. Let me know if you need further help to get it running. Glad to hop on a call :)

Aside from testing and QA, is there anything else we need to get this merged and published?

File: server/utils/textTracks.ts (hunk 2 above)
Comment (hdoro) [reply server/utils/textTracks.ts]: This wasn't being used, actually - it was a leftover from an initial iteration

File: package.json [hunk 4]
```diff
@@ -23,43 +23,47 @@
...
+    "formik": "^2.4.5",
+    "iso-639-1": "^3.1.0",
     "lodash": "^4.17.21",
-    "luxon": "^3.3.0",
-    "react": "^17.0.2",
-    "react-dom": "^17.0.2",
-    "react-intl": "5.20.2",
+    "luxon": "^3.4.4",
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
```
[the remark this answers is not in the document]
Comment (hdoro) [reply package.json:45]: Yup! I've seen issues only when updating react-router-dom, styled-components or react-intl. `react` and `react-dom` are fine. Should I undo this upgrade, though?

[the remark this answers is not in the document]
Comment (erikpena) [reply package.json:45]: I think that if we can still get away with using v17 for these libraries and (more importantly) it still works, we should.  We can fight this battle another day ;-)

**Comment by erikpena** [comment]
Hello @hdoro,
I noticed that you had moved the delete action button for assets.  We will need it where it was at as this was based on conditions from Strapi in order for us to get approved and "Verified".

**Comment by erikpena** [comment]
Hey @hdoro I just attempted to create a new subtitle with an asset and it's throwing a 500.

I used a file like this (named `test.srt`)—

```
1
00:00:00,498 --> 00:00:02,827
- Here's what I love most
about food and diet.

2
00:00:02,827 --> 00:00:06,383
We all eat several times a day,
and we're totally in charge
```

Selected "English"  and left "closed captions" unchecked.  When submitting the form, I get this error in the console for my —

```
[updateTextTracks / createAssetTextTracks] BadRequestError: 400 {"error":{"type":"invalid_parameters","messages":["input has url set to \"undefined/mux-video-uploader/mux-text-tracks/2\", which is not a valid URL."]}}
```

In my case, I'm hosting locally so I'm not expecting it to work fully.  However, it looks like it's having issues with the url resolution on the input when sent to Mux (there is an undefined value).  Can you determine what is causing this?

**Comment by erikpena** [comment]
I can see that there is still a file that is camel cased (server/utils/parseJSONBody.ts), can we change this to kebab casing (server/utils/parse-json-body.ts)?

**Review by erikpena — APPROVED** [review]
LGTM!

File: server/utils/text-tracks.ts [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+    type: 'text',
+    text_type: 'subtitles',
+    closed_captions: track.closed_captions,
+    language_code: track.language_code,
+    name: track.closed_captions ? `${track.name} (CC)` : track.name,
+    url: getStrapiTextTrackUrl(track.id),
+  };
+}
+
+const BASE_URL = (strapi as any).backendURL;
```
Comment (kelihansen) [inline server/utils/text-tracks.ts:30]: Is this available on the server @hdoro? Like @erikpena, I'm getting
```
[updateTextTracks / createAssetTextTracks] BadRequestError: 400 {"error":{"type":"invalid_parameters","messages":["input has url set to \"undefined/mux-video-uploader/mux-text-tracks/3\", which is not a valid URL."]}}
```

**Pull request merged**